include/yk/colorize.hpp

```diff
@@ -0,0 +1,803 @@
+#ifndef YK_COLORIZE_HPP
+#define YK_COLORIZE_HPP
+
+#include "yk/detail/string_like.hpp"
+#include "yk/enum_bitops.hpp"
+#include "yk/enum_bitops_algorithm.hpp"
+
+#include <algorithm>
+#include <array>
+#include <format>
+#include <optional>
+#include <ranges>
+#include <stdexcept>
+#include <string_view>
+#include <utility>
+#include <version>
+
+#include <cstdint>
+
+namespace yk {
+
+class colorize_error : public std::runtime_error {
+  using runtime_error::runtime_error;
+};
+
+template <class CharT>
+struct basic_colorize_parse_context {
+  using char_type = CharT;
+  using const_iterator = typename std::basic_string_view<CharT>::const_iterator;
+  using iterator = const_iterator;
+
+  constexpr explicit basic_colorize_parse_context(std::basic_string_view<CharT> fmt) noexcept
+      : begin_(fmt.begin()), end_(fmt.end())
+  {
+  }
+
+  constexpr iterator begin() const noexcept { return begin_; }
+  constexpr iterator end() const noexcept { return end_; }
+  constexpr void advance_to(iterator it) noexcept { begin_ = it; }
+
+private:
+  iterator begin_;
+  iterator end_;
+};
+
+template <class Out, class CharT>
+struct basic_colorize_context {
+  using char_type = CharT;
+  using iterator = Out;
+
+  explicit constexpr basic_colorize_context(Out out) : out_(std::move(out)) {}
+
+  constexpr iterator out() { return std::move(out_); }
+  constexpr void advance_to(iterator it) { out_ = std::move(it); }
+
+private:
+  iterator out_;
+};
+
+using colorize_parse_context = basic_colorize_parse_context<char>;
+
+namespace detail {
+
+// https://github.com/fmtlib/fmt/blob/master/include/fmt/color.h
+enum class rgb_color : uint32_t {
```

Comment:
According to my experience as a designer, it feels more natural if the color scheme is grouped like this:

- pure white
- ... gray colors ...
-  pure black
- (empty line)
- ... [pure colors](https://ja.wikipedia.org/wiki/%E7%B4%94%E8%89%B2) ...
- (empty line)
- other colors